File: java/ApproxPiReporter.java
import java.io.*;
import java.util.Formatter;

class ApproxPiReporter
{
    private static int REPEATS = 1000;

    private int sampleSize;
    private double lowestError;
    private double highestError;

    private double bestPi;
    private double worstPi;
    private double sumOfPis = 0.0;

    private void generate() {
        ApproxPiFinder finder = new ApproxPiFinder(sampleSize);
        finder.evaluate();
        double approxPi = finder.approxPi();
        double error = finder.error();

        if (Math.abs(error) < Math.abs(lowestError))
        {
            bestPi = approxPi;
            lowestError = error;
        }

        if (Math.abs(error) > Math.abs(highestError))
        {
            worstPi = approxPi;
            highestError = error;
        }

        sumOfPis += approxPi;
    }

    private void repeatGenerates() {
        lowestError = 100.0;
        highestError = 0.0;
        sumOfPis = 0.0;

        for (int i = 0; i < REPEATS; i++)
            generate();
    }

    private void outputToConsole() {
        System.out.println("\n");
        System.out.println("For sample size of " + sampleSize);
        System.out.println("Best PI  : " + bestPi + "(error = " + lowestError + "%)");
        System.out.println("Worst PI : " + worstPi + "(error = " + highestError + "%)");
        System.out.println("Avg. PI  : " + averagePi() + "(error = " + averageError() + "%)");
    }

    private double averagePi() {
        return sumOfPis / REPEATS;
    }
 
    private double averageError() {
        return (averagePi() - Utils.actualPi()) * 100.0 / Utils.actualPi();
    }

    private void outputToFiles() {
        writeToFile("../output/java.average.csv", averageError());
        writeToFile("../output/java.best.csv", lowestError);
        writeToFile("../output/java.worst.csv", highestError);
    }

    private void writeToFile(String fileName, double value) {
        try {
            FileWriter fstream = new FileWriter(fileName, true);
            Formatter format = new Formatter(fstream);
            format.format("%d,%.4f%n", sampleSize, value);
            format.flush();
            format.close();
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    public ApproxPiReporter(int sampleSize) {
        this.sampleSize = sampleSize;
    }

    public void output() {
        repeatGenerates();
        outputToConsole();
        outputToFiles();
    }
}

File: go/approx_pi_reporter.go
package main

import (
	"fmt"
	"io"
	"math"
	"os"
)

const (
	repeats = 1000
)

type ApproxPiReporter struct {
	sampleSize   int
	lowestError  float64
	highestError float64

	bestPi   float64
	worstPi  float64
	sumOfPis float64
}

func newApproxPiReporter(sampleSize int) ApproxPiReporter {
	return ApproxPiReporter{
		sampleSize: sampleSize,
		sumOfPis:   0.0,
	}
}

func (reporter *ApproxPiReporter) generate() {
	finder := newApproxPiFinder(reporter.sampleSize)
	finder.evaluate()
	approxPi := finder.approxPi()
	error := finder.error()

	if math.Abs(error) <= math.Abs(reporter.lowestError) {
		reporter.bestPi = approxPi
		reporter.lowestError = error
	}

	if math.Abs(error) >= math.Abs(reporter.highestError) {
		reporter.worstPi = approxPi
		reporter.highestError = error
	}

	reporter.sumOfPis += approxPi
}

func (reporter *ApproxPiReporter) repeatGenerates() {
	reporter.lowestError = 100.0
	reporter.highestError = 0.0
	reporter.sumOfPis = 0.0

	for i := 0; i < repeats; i++ {
		reporter.generate()
	}
}

func (reporter *ApproxPiReporter) outputToConsole() {
	fmt.Println("\n")
	fmt.Printf("For sample size of %d\n", reporter.sampleSize)
	fmt.Printf("Best PI  : %f (error = %f%%)\n", reporter.bestPi, reporter.lowestError)
	fmt.Printf("Worst PI : %f (error = %f%%)\n", reporter.worstPi, reporter.highestError)
	fmt.Printf("Avg. PI  : %f (error = %f%%)\n", reporter.averagePi(), reporter.averageError())
}

func (reporter *ApproxPiReporter) averagePi() float64 {
	return reporter.sumOfPis / repeats
}

func (reporter *ApproxPiReporter) averageError() float64 {
	return (reporter.averagePi() - math.Pi) * 100.0 / math.Pi
}

func (reporter *ApproxPiReporter) outputToFiles() {
	reporter.writeToFile("../output/go.average.csv", reporter.averageError())
	reporter.writeToFile("../output/go.best.csv", reporter.lowestError)
	reporter.writeToFile("../output/go.worst.csv", reporter.highestError)
}

func (reporter *ApproxPiReporter) writeToFile(fileName string, value float64) {
	text := fmt.Sprintf("%d,%.4f\n", reporter.sampleSize, value)

	f, _ := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE, 0666)
	io.WriteString(f, text)
	f.Close()
}

func (reporter ApproxPiReporter) output() {
	reporter.repeatGenerates()
	reporter.outputToConsole()
	reporter.outputToFiles()
}
